### PR TITLE
[INTEL MKL] Added support for Quantized-Conv2D and Pad fusion.

### PR DESCRIPTION
**Pull request by bhavani-subramanian**

### CODE REVIEW

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (shahzadlone) [inline tensorflow/core/kernels/mkl_conv_ops.cc]: I know you didn't do this but could it change this ```i++``` to ```++i``` while at it, please..

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you very much for the PR!

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Since `padding_list` is of type `vector<int32>` in the layout pass, `Tpaddings` needs to be `int32` as well - otherwise the compiler complains about not being able to convert `int*` to `long long int*`  in line 1167:
```
1166 if (quantized_pad_enabled) {
1167  paddings = padding_list_.data();
1168} else {
```

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Same comment as above.

File: tensorflow/core/kernels/mkl_quantized_conv_ops_test.cc
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_quantized_conv_ops_test.cc]: Good catch! It should be `qint32`.

File: tensorflow/core/kernels/mkl_quantized_conv_ops_test.cc
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_quantized_conv_ops_test.cc]: Done.

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Done. Thanks!

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Agree. Changed it per your suggestion.

File: tensorflow/core/kernels/mkl_quantized_conv_ops_test.cc
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_quantized_conv_ops_test.cc]: Done.

**Comment by bhavani-subramanian** [comment]
@penpornk Thanks for quickly reviewing the PR! I have addressed your review comments. Let me know if you need anything else.

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (penpornk) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: But then why do you need the type constraint 3 lines above to be `int64`? 
https://github.com/tensorflow/tensorflow/blob/0a507dc55b51e4f5651b57473ff5e2b3ea50d95e/tensorflow/core/kernels/mkl_conv_ops.cc#L2090

**Review by penpornk — COMMENTED** [review]
Thank you for your quick changes!

**Comment by bhavani-subramanian** [comment]
@penpornk I have changed `padding_list_` to be of type `Tpadding` instead of `int32`.

Also, I have added an additional registration for when `Tbias` is a `float` since we support both `qint32` and `float` types for Tbias for `Quantized-Conv2D + bias + sum + relu + requantize` fusion.

**Review by penpornk — APPROVED** [review]
Looks good. Thank you again for the PR! :)

**Comment by bhavani-subramanian** [comment]
@pragyaak The unit test failures don't seem to be related to the changes made in this PR. Please let me know if you think otherwise.

**Review by penpornk — CHANGES_REQUESTED** [review]
@bhavani-subramanian The `Ubuntu Python2` and `Ubuntu Python3` failures are actually related to this PR. The API golden files need to be modified because we added an extra attribute `padding_list` to the ops. See the `Ubuntu Python 2` test log [here](https://source.cloud.google.com/results/invocations/be1124ab-1bf9-4542-b3c8-f61de7af5a4b/targets/%2F%2Ftensorflow%2Ftools%2Fapi%2Ftests:api_compatibility_test/log).

Could you please run the following commands and add the modified golden files (`tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt`, and `tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt`)to the PR?
```
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```
After doing this, your local `bazel test tensorflow/tools/api/tests:api_compatibility_test` should pass. 
Thank you!

**Comment by bhavani-subramanian** [comment]
Thanks for the explanation, @penpornk! I tried modifying the API golden files per your suggestion, but only `tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt` ended up getting updated, and `tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt` remains unmodified. However, I do see that `bazel test tensorflow/tools/api/tests:api_compatibility_test` unit test passes locally even with just this change.

**Comment by penpornk** [comment]
@bhavani-subramanian We pulled a PR in from last week with just changes to `v2`, but it failed the internal test for `v1` and we had to add changes to `v1` anyway. So I think we should add it now. I'm not sure how to get the OSS test to modify `v1`, though. (Apart from modifying it manually from the diff.) 

I'm asking @annarev: 
Do you know how to make the api_compatibility_test also modify the `v1` file? Thank you!

**Comment by annarev** [comment]
I submitted a fix for that issue last week. @bhavani-subramanian can you sync to head and check if api_compatibility_test updates both v1 and v2 files now?

**Comment by bhavani-subramanian** [comment]
Thanks for the quick response, @annarev. I was able to add the `v1` file after merging my branch with the master.

@penpornk Please let me know if anything else is needed from my end. Thanks!

**Review by penpornk — APPROVED** [review]
@bhavani-subramanian I think it looks good now. Thank you so much for the changes! (And thank you, @annarev!)

Re-running the tests and waiting for API review then.

**Comment by bhavani-subramanian** [comment]
@penpornk The two failures in Ubuntu contrib pass when run locally. Not sure how to reproduce the error at my end.

**Comment by penpornk** [comment]
@bhavani-subramanian I'll rerun the tests. I can't see the log of `Ubuntu Python 2`. The rest looks unrelated.

**Comment by bhavani-subramanian** [comment]
@penpornk Sorry to bug you again. The unit test that fails in `MacOS Python2 and CC` (`//tensorflow/python/saved_model:load_test`) passes when run locally. From the error log, it looks like the test failed because it was unable to create a new thread either because the system either ran out of resources (or) due to hitting a system limit on the number of threads per process ([Reference](https://stackoverflow.com/questions/26691340/multi-threading-in-c-throws-thread-constructor-failed-resource-temporarily-un)).

It doesn't seem like this failure is related to my PR.

**Comment by penpornk** [comment]
@bhavani-subramanian Ah. Don't worry about them! All three failing tests right now are existing failures (`MacOS Python2 and CC`, `Windows Bazel`, and `Windows Bazel GPU`) I mainly reran the tests because `Ubuntu Python 2` and `Ubuntu Contrib` shouldn't be failing. They are passing now so I think we are fine.

I haven't tagged `ready to pull` yet because this still needs API review (because we changed the golden files).

**Comment by Srini511** [comment]
@pragyaak gentle reminder regarding the API review for this PR.

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: @penpornk Just for my learning purposes, can you help me understand how having a `continue` results in more jumps? Any pointers to material would also help. Thanks!

**Comment by penpornk** [comment]
Apologies for the delay, but this PR finally got through the API review and is getting merged internally. But we got outdated files, specifically,
```
//depot/google3/third_party/tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt
//depot/google3/third_party/tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt
//depot/google3/third_party/tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt
```
Could you please help resolve these? Thank you!

**Comment by bhavani-subramanian** [comment]
@penpornk I merged my branch with the master and ran the commands below. But the above files are still not getting updated:
```
 $ bazel build tensorflow/tools/api/tests:api_compatibility_test
 $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```
Are there other commands I should be running instead?

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (penpornk) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: Sorry about the confusion! The codes should have the same number of jump instructions. I was just thinking if one might need to jump less based on what values padding_val usually have. I should have added "probably" because I didn't try to count properly and was just speculating what padding_list_ would be like.
```
  compare_condition
  jump_to_"false_label"_if_comparison_is_false
  do_something_if_true
false_label:
  commands_outside_if
```

**Comment by penpornk** [comment]
@bhavani-subramanian Do you have the changes from the latest commit on these two files?: https://github.com/tensorflow/tensorflow/commit/4294edea3e80c746eac400262628f852b66d02e5

**Comment by penpornk** [comment]
@bhavani-subramanian Never mind! It's getting merged now. :)

**Comment by bhavani-subramanian** [comment]
Thanks @penpornk! I am somehow not able to generate the golden files despite having the `4294ede` commit. Anyway, glad that its getting merged now :)

File: tensorflow/core/kernels/mkl_conv_ops.cc
Comment (bhavani-subramanian) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: No worries. Thanks for the explanation!

**Pull request merged**
